components/style/values/specified/grid.rs

```diff
@@ -2,73 +2,19 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-//! Necessary types for [grid](https://drafts.csswg.org/css-grid/).
+//! CSS handling for the computed value of
+//! [grids](https://drafts.csswg.org/css-grid/)
 
-use cssparser::{Parser, Token, serialize_identifier};
+use cssparser::{Parser, Token};
 use parser::{Parse, ParserContext};
-use std::{fmt, mem, usize};
+use std::{mem, usize};
 use std::ascii::AsciiExt;
-use style_traits::ToCss;
 use values::{CSSFloat, CustomIdent, Either, HasViewportPercentage};
-use values::computed::{self, ComputedValueAsSpecified, Context, ToComputedValue};
+use values::computed::{self, Context, ToComputedValue};
+use values::generics::grid::{GridLine, TrackBreadth, TrackKeyword, TrackSize};
+use values::generics::grid::{RepeatCount, TrackRepeat, TrackList, TrackListType};
 use values::specified::{Integer, LengthOrPercentage};
 
-#[derive(PartialEq, Clone, Debug)]
-#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
-/// A `<grid-line>` type.
-///
-/// https://drafts.csswg.org/css-grid/#typedef-grid-row-start-grid-line
-pub struct GridLine {
-    /// Flag to check whether it's a `span` keyword.
-    pub is_span: bool,
-    /// A custom identifier for named lines.
-    ///
-    /// https://drafts.csswg.org/css-grid/#grid-placement-slot
-    pub ident: Option<String>,
-    /// Denotes the nth grid line from grid item's placement.
-    pub line_num: Option<Integer>,
-}
-
-impl GridLine {
-    /// Check whether this `<grid-line>` represents an `auto` value.
-    pub fn is_auto(&self) -> bool {
-        self.ident.is_none() && self.line_num.is_none() && !self.is_span
-    }
-}
-
-impl Default for GridLine {
-    fn default() -> Self {
-        GridLine {
-            is_span: false,
-            ident: None,
-            line_num: None,
-        }
-    }
-}
-
-impl ToCss for GridLine {
-    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
-        if self.is_auto() {
-            return dest.write_str("auto")
-        }
-
-        if self.is_span {
-            dest.write_str("span")?;
-        }
-
-        if let Some(i) = self.line_num {
-            write!(dest, " {}", i.value)?;
-        }
-
-        if let Some(ref s) = self.ident {
-            dest.write_str(" ")?;
-            serialize_identifier(s, dest)?;
-        }
-
-        Ok(())
-    }
-}
-
 impl Parse for GridLine {
```

Comment:
Let's also move this `Parse` impl to generics, since the specified and computed values have the same type anyway.